sdk/identity/azure-identity/src/client_assertion_credential.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "azure/identity/client_assertion_credential.hpp"
+
+#include "private/identity_log.hpp"
+#include "private/package_version.hpp"
+#include "private/tenant_id_resolver.hpp"
+#include "private/token_credential_impl.hpp"
+
+#include <azure/core/internal/json/json.hpp>
+
+using Azure::Identity::ClientAssertionCredential;
+using Azure::Identity::ClientAssertionCredentialOptions;
+
+using Azure::Core::Context;
+using Azure::Core::Url;
+using Azure::Core::_internal::StringExtensions;
+using Azure::Core::Credentials::AccessToken;
+using Azure::Core::Credentials::AuthenticationException;
+using Azure::Core::Credentials::TokenRequestContext;
+using Azure::Core::Http::HttpMethod;
+using Azure::Identity::_detail::IdentityLog;
+using Azure::Identity::_detail::TenantIdResolver;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+bool IsValidTenantId(std::string const& tenantId)
```

Comment:
I noticed we have this in AzurePipelinesCredential also - is it possible to centralize it to a single implementation?

Reply:
`AzureCliCredential` has almost identical logic too, maybe that one can also be centralized?

Reply:
> I noticed we have this in AzurePipelinesCredential also - is it possible to centralize it to a single implementation?

Good question. I will be moving most of the APC credential implementation (and WIC impl) to leverage ClientAssertionCredential, as a **follow-up** to this PR. That will remove the duplication you see.

The usage pattern in AzureCliCredential is a bit different (for example: different method semantics, throw rather than return false, empty string isn't considered invalid, and it re-uses that method validation between scopes/tenant id).

If we think it's worth consolidating between the two, using a utility validation method, I will do that as a separate PR to avoid touching other existing credential types in a PR introducing a new one.

Reply:
Resolved in https://github.com/Azure/azure-sdk-for-cpp/pull/5802#discussion_r1678797071